### PR TITLE
Proposal: Sweepable API

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
This is the initial design/proposal doc for our cross platform approach.

I would appreciate it if you could each review it and give me your feedback.

### CODE REVIEW

File: docs/specs/ML.Net Sweepable API.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
+# AutoML.Net Sweepable API proposal
```
[the remark this answers is not in the document]
Comment (JakeRadMSFT) [reply docs/specs/ML.Net Sweepable API.md:1]: @jwood803  ( I saw you're like ) we'd love to hear your feedback as well.

[the remark this answers is not in the document]
Comment (jwood803) [reply docs/specs/ML.Net Sweepable API.md:1]: Hey, @JakeRadMSFT. I definitely like having this API! This helps make it more on part with scikit-learn that has grid search. Plus, I believe this will help model creators get better models.

[the remark this answers is not in the document]
Comment (eerhardt) [reply docs/specs/ML.Net Sweepable API.md:1]: > I'm open to keeping these layers internal if we need to

My suggestion would be to start with them internal, and ensure it can meet the existing AutoML scenarios. Then make them public as we go when we have data that supports they are needed to be public.

> The end goal is to only have one ML.NET AutoML.

💯

File: docs/specs/ML.Net Sweepable API.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+Sweepable api allows mlnet users to create their own search space and pipeline for hyper-parameter optimization (HPO). It comes with three major part: [search space](#search-space), [sweepable estimator](#sweepable-estimator) and [tuner](#tuner). And all API lives under `Sweepable()` extension (for now).
+
+## search space
+Search space defines a range of hyper-parameter for tuner to search from. Sweepable API provides two way to create a search space.
+
+via attribute
+```csharp
+public class Option
+{
+    [Range(2, 32768, init: 2, logBase: true)]
```
Comment (eerhardt) [inline docs/specs/ML.Net Sweepable API.md:12]: Do we need to have a reflection based solution to start? Or would the "weakly-typed" API solve most of the scenarios:

```C#
var ss = new SearchSpace();
ss.Add("WindowSize", new UniformIntOption(2, 32768, true, 2));
ss.Add("SeriesLength", new ChoiceOption(2,3,4));
ss.Add("UseSoftmax", new ChoiceOption(true, false));
ss.Add("AnotherOption", ss.Clone())
```

Comment (eerhardt) [reply docs/specs/ML.Net Sweepable API.md:12]: The reason I ask is: I would not add multiple ways to do something right away. Instead, make the "core" thing first, and the simpler ones can be built later, if necessary.

Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:12]: The reflection API (let's call it strong-typed correspond with "weakly-typed") is just a handy way for creating search space and it's built on top of "weakly-typed". So if we must make a choice at the beginning I would pick "weakly-typed" to implement/migrate first.

File: docs/specs/ML.Net Sweepable API.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:1]: Sounds good, if we want to start from internal first, a smooth start without breaking change can be search space, either through package reference or source code. The reasons are
- search space has the least dependency. It only depends on newtonsoft, which makes it easy to introduce.
- after introducing, we can use it to create search space for existing trainers in AutoML.Net. Which provides us a chance to improve the performance of AutoML.Net experiments by using larger, better search space in hpo as well.

[the remark this answers is not in the document]
Comment (justinormont) [reply docs/specs/ML.Net Sweepable API.md:1]: Sounds like a fine plan, including keeping as internal for a bit.

If it fits within your methods, I'd recommend having the _default_ sweeping space { range, scaling, data type } as part of the component being swept (as [current](https://github.com/dotnet/machinelearning/blob/510f0112d4fbb4d3ee233b9ca95c83fae1f9da91/src/Microsoft.ML.FastTree/FastTreeArguments.cs#L508-L509) params). For AutoML․NET, due to non-great reasons, we kept a duplicate copy. Traveling with the component is more clean since the ranges show up (and disappear) with the availability of the component. For instance, as the user includes the NuGet for FastTree, the ranges are immediately available. In addition the ranges can be created inline with the new component. 

Some historical perspective, there is also a command-line sweeper in ML․NET, which may be working (anything untested is assumed non-functional). This does let users sweep an individual (or sets of) components including fully switching out portions of the pipeline. The MAML command-line sweeper was very powerful, but hard to use. Mentioned previously: https://github.com/dotnet/machinelearning/pull/5019#issuecomment-817105977

[the remark this answers is not in the document]
Comment (eerhardt) [reply docs/specs/ML.Net Sweepable API.md:1]: > It only depends on newtonsoft

If possible, can we use System.Text.Json?

File: docs/specs/ML.Net Sweepable API.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+sweepable estimator allows user to combine search space with estimators in a similar way how ml.net estimator/pipeline are created. You use `CreateSweepableEstimator`, which accepts a lambda function and a search space to create a sweepable estimator. And it also provides `.Append` extension method so you can append sweepable estimator similiar with how you append other ml.net extimator. The bellow example presents how to create a pipeline with two sweepable estimators, one for text featurizor and the other for fast tree, for `titanic` dataset via `Sweepable()` extension.
+
+``` csharp
+var context = new MLContext();
+var fastTreeSS = new SearchSpace<FastTreeOption>();
+var textFeaturizeSS = new SearchSpace<FeaturizeTextOption>();
+
+var pipeline = context.Transforms.Categorical.OneHotEncoding(new[] { new InputOutputColumnPair(@"Sex", @"Sex"), new InputOutputColumnPair(@"Embarked", @"Embarked") })
+                .Append(context.Transforms.Concatenate(@"TextFeature", @"Name", "Ticket", "Cabin"))
+                .Append(context.Sweepable().CreateSweepableEstimator(
```
Comment (michaelgsharp) [inline docs/specs/ML.Net Sweepable API.md:75]: Is there a reason the `Sweepable()` extension uses the parenthesis when none of the other extensions on the MLContext use them?

Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:75]: `Sweepable()` is not the only extension that uses parenthesis. AFAIK `Auto()` and `Recommendation` extension also have parenthesis.

File: docs/specs/ML.Net Sweepable API.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                    textFeaturizeSS))
+                .Append(context.Transforms.Concatenate(@"Features", new[] { @"Sex", @"Embarked", @"Pclass", @"Age", @"SibSp", @"Parch", @"Fare", "TextFeature" }))
+                .Append(context.Transforms.Conversion.ConvertType("Survived", "Survived", Data.DataKind.Boolean))
+                .Append(context.Sweepable().CreateSweepableEstimator(
+                    (mlContext, option) => mlContext.BinaryClassification.Trainers.FastForest(labelColumnName: "Survived", featureColumnName: "Features", numberOfLeaves: option.NumberOfLeavenumberOfTrees: option.NumberOfTrees),
+                    fastTreeSS))
+                .Append(context.BinaryClassification.Calibrators.Naive(labelColumnName: @"Survived", scoreColumnName: @"Score"));
+```
+
+After sweepable pipeline is created, one can call `BuildTrainingPipeline` to convert it to a ml.net pipeline.
```
Comment (michaelgsharp) [inline docs/specs/ML.Net Sweepable API.md:103]: Will the type of the pipeline be the same as a normal ML.NET pipeline? What happens if someone tries to run it without calling that method first?

File: docs/specs/ML.Net Sweepable API.md (hunk 3 above)
Comment (michaelgsharp) [reply docs/specs/ML.Net Sweepable API.md:75]: Ah yeah, I see. Its based on if its an extension method or not.

File: docs/specs/ML.Net Sweepable API.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+var df = DataFrame.LoadCsv(@"titanic.csv");
+var trainTestSplit = context.Data.TrainTestSplit(df, 0.1);
+var bestAccuracy = 0.0;
+var i = 0;
+foreach (var param in tuner.Propose())
+{
+    Console.WriteLine($"trial {i++}");
+
+    // convert sweepable pipeline to ml.net pipeline
+    var trainingPipeline = pipeline.BuildTrainingPipeline(context, param);
```
Comment (michaelgsharp) [inline docs/specs/ML.Net Sweepable API.md:123]: Just curious, but would there ever be a need to save the current state of the sweeper so that it can be stopped and then resumed at a later date?

File: docs/specs/ML.Net Sweepable API.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+What's the timeline for intergrating Sweepable API into ML.Net
+- it depends. Sweepable API has three major parts and all are available in model builder repo. On top of which is a thin layer which makes the three part work with ML.Net. So if the plan is only moving that thin layer into AutoML.Net first, and make the rest of three parts open source later, it will take 2-3 weeks. If we want to move all parts into mlnet all together, it will cost much more time.
+
+
+## Current feedback
+- the lambda function in `CreateSweepableEstimator` should not accept `MLContext` as first parameter.
+- Should provide a higher level API for training similar with Experiment API in AutoML.Net.
```
Comment (michaelgsharp) [inline docs/specs/ML.Net Sweepable API.md:162]: Should that higher level API be part of this though?  Wouldn't that belong more in the AutoML side and less with the sweeper side since the sweeper is "lower level" and is expected to be more hands on?

File: docs/specs/ML.Net Sweepable API.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:1]: > If possible, can we use System.Text.Json?

Currently no because auto-binding experience depends heavily on `JToken`. However, there's a plan to remove this dependency so in the future maybe.

File: docs/specs/ML.Net Sweepable API.md (hunk 4 above)
Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:103]: No it won't be the same ML.Net pipeline. And user needs to call `BuildTrainingPipeline` with passed parameter to convert it to an ML.Net pipeline and only after that, they can use `Fit/Transform` to train their model.

I feel like that `BuildTrainingPipeline` might be too detailed an API to be exposed. And it will be better if it can be wrapped and only exposes a `Run` method. We can discuss it later though.

File: docs/specs/ML.Net Sweepable API.md (hunk 5 above)
Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:123]: If sweeper supports that then yes. In fact continue training has already been supported in model builder cfo tuner, it's just not exposed yet.

File: docs/specs/ML.Net Sweepable API.md (hunk 6 above)
Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:162]: Yup the higher level API will be along with exsiting AutoML Experiment API.

File: docs/specs/ML.Net Sweepable API.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:1]: ### Resolved

File: docs/specs/ML.Net Sweepable API.md (hunk 2 above)
Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:12]: ### Resolved

**Comment by torronen** [comment]
From user perspective this looks great and promising! As I understand pipeline.SearchSpace will automatically keep track of the experiments, their results and parameters. So, basically, we could start each new tuning experiment by first running the best parameters from last experiment, then looping with tuner.Propose. Is it correct? 

If possible, I'd also like to extract the results from pipeline.SearchSpace to help me get a better understanding of how the tuning affects performance. For example, maybe plot a graph of how the number of trees affect FastForest to help me find optimal number considering training time vs. accuracy.

File: docs/specs/ML.Net Sweepable API.md (hunk 5 above)
Comment (torronen) [reply docs/specs/ML.Net Sweepable API.md:123]: Maybe for the future, but how about "import experiments"? That would help in case someone wants to distribute training to multiple nodes: a) save each experiment results to db  b) before starting next iteration, fetch new from db and import

Comment (LittleLittleCloud) [reply docs/specs/ML.Net Sweepable API.md:123]: @torronen 
That's a wonderful suggestion, thanks!

**Comment by LittleLittleCloud** [comment]
> From user perspective this looks great and promising! As I understand pipeline.SearchSpace will automatically keep track of the experiments, their results and parameters. So, basically, we could start each new tuning experiment by first running the best parameters from last experiment, then looping with tuner.Propose. Is it correct?

@torronen Quite close but not exactly accurate. Pipeline.SearchSpace will only define the boundary of parameters and and it's stateless. If you want to track parameter and their results under current api, you need to do it yourself. (save each parameter and their result in a dictionary or something..)

> If possible, I'd also like to extract the results from pipeline.SearchSpace to help me get a better understanding of how the tuning affects performance. For example, I might want to plot a graph about how the number of trees affect FastForest to help me find optimal number considering training time vs. accuracy.

That's a great advice. It's also something that we want to heavily dive into on how to help users understand training process, like trend line of metric, importance of each parameter, etc etc...

**Comment by torronen** [comment]
@LittleLittleCloud ok I see. I assume Tuner needs to keep the results. 

Not sure if these are valid for this plan, but my next questions might be:
- How to change the optimization metric? 
- How to create custom optimization metric? See https://github.com/dotnet/machinelearning/issues/5999 Maybe derived from other metrics (maybe as anonymous method?), or even create a mini-simulation. Based on my initial experiments with simulating real-life events, simulation results might be more sensitive to tuning than traditional metrics. For example, we may want to catch some of different types of malicious web requests to catch the bad actors. Same ML metrics may come for detecting 100% of all requests for a single type, but in this case we might only catch one bad actor. In a simulation we might want to see how early intruders are detected: it's not enough to detect intruder when the logout.

- How to create a new Tuner? Not sure if necessary. It can also be done without being this type of Tuner. Could be also one area where to make it easy for community to contribute their implementations as extension nugets, or even as part of Microsoft.ML if dev team thinks it is feasible.

**Comment by LittleLittleCloud** [comment]
@torronen  reply inline

**Comment by torronen** [comment]
@LittleLittleCloud BTW, does this proposal and search space API allow optimizing the booster in LightGBM? Or, should users run the tuning for DART, Gradient and Goss separately is they expect it to matter? Asking because it is an object, not value type, and the object has parameters inside it. 

Or, maybe user can write his own if...else or switch depending on what the tuner wants to run next?

**Comment by LittleLittleCloud** [comment]
> @LittleLittleCloud BTW, does this proposal and search space API allow optimizing the booster in LightGBM? Or, should users run the tuning for DART, Gradient and Goss separately is they expect it to matter? Asking because it is an object, not value type, and the object has parameters inside it.
> 
> Or, maybe user can write his own if...else or switch depending on what the tuner wants to run next?

Yes you can, it really doesn't matter whether the property you want to sweep on is a value type or not, you can aways pass a factory method to create the trainer based on parameter sweeped from search space.

**Comment by michaelgsharp** [comment]
@LittleLittleCloud Merging this in as the sweepable estimator itself is already checked in.
